### PR TITLE
Support for serialization of TR_BlockFrequencyInfo

### PR DESCRIPTION
**Pull request by ashu-mehra**
Added code for serialization of TR_BlockFrequencyInfo.

Signed-off-by: Ashutosh Mehra <asmehra@redhat.com>

### CODE REVIEW

**Comment by ashu-mehra** [comment]
This PR depends on https://github.com/eclipse/omr/pull/5201 for serialization of `TR_BitVector` and `TR_Serializer` class.

**Comment by ashu-mehra** [comment]
fyi - @mpirvu @andrewcraik @r30shah @dsouzai

**Comment by ashu-mehra** [comment]
These changes can't be tested right now as the code path for serialization won't be exercised.
I will be working on enabling JProfiling on JITServer which would allow to test these changes.

File: runtime/compiler/runtime/J9Profiler.hpp
Comment (andrewcraik) [inline runtime/compiler/runtime/J9Profiler.hpp:437]: Can you explain why this is ifdef'd out? Ordinarily we wouldn't accept ifdef'd out code since it is likely to rot and isn't tested or compiled....

File: runtime/compiler/runtime/J9Profiler.hpp
Comment (andrewcraik) [inline runtime/compiler/runtime/J9Profiler.hpp:483]: same as above ifdef

File: runtime/compiler/runtime/J9Profiler.hpp
Comment (andrewcraik) [inline runtime/compiler/runtime/J9Profiler.hpp:180]: Making this public seems suboptimal - the isolation provided by the previous private declaration helps to prevent accidental corruption of the profiling data. Is there an overall design for this work that explains this design choice?

File: runtime/compiler/runtime/J9Profiler.hpp
Comment (andrewcraik) [inline runtime/compiler/runtime/J9Profiler.hpp:181]: Can you please add doxygen for newly added APIs to explain what they do, their requirements, results and side-effects?

**Comment by dsouzai** [comment]
> Is there a design for this at the OMR level? 

Could you also open an issue on the OpenJ9 side outlining the general idea behind the AOT support for JProfiling, both the Block Frequency work you have here as well as the eventual Value Profiling work to come (since https://github.com/eclipse/openj9/issues/4848 doesn't really have any design information).

**Comment by ashu-mehra** [comment]
@andrewcraik I have updated the PR as per the suggestions earlier and also added some more changes. This is ready to review now.

**Comment by ashu-mehra** [comment]
@r30shah @dsouzai @mpirvu  this PR is ready to be reviewed. Its omr dependency is also merged.

**Comment by r30shah** [comment]
@ashu-mehra  Is there a reason why commits are separated? I.e, looking at https://github.com/eclipse/openj9/pull/9590/commits/4920d979884e321fec8d5d5f0a48dfd66d7e7bae and https://github.com/eclipse/openj9/pull/9590/commits/5e880443328e9fa44d547be81aa56f9ea7ae7f19, I see some of the changes from https://github.com/eclipse/openj9/pull/9590/commits/482fa54d5e812439945c80e506a1c3683a41f947 were changed.

**Comment by ashu-mehra** [comment]
Initial commit used `TR_Serializer` which is no longer present. Second commit 4920d97 removed the dependency on `TR_Serializer`. Last commit 5e88044 is some refactoring in the way serialization of fields is managed to be consistent with the approach used in https://github.com/eclipse/omr/pull/5693. 
These commits can be squashed into single one once the changes are approved.

File: runtime/compiler/runtime/J9Profiler.hpp
Comment (r30shah) [inline runtime/compiler/runtime/J9Profiler.hpp:667]: I do not think so `IS_VALID_BIT_VECTOR` is the correct name here. 
IIRC low tagged entry means we need to just get the counter from single block to either add or subtract, otherwise, it should hold the array of bit vector representing counters from blocks added and subtracted together to get the frequency. I am not sure if this is the correct check. Take a look at the comment here,
https://github.com/eclipse/openj9/blob/b2400f9240b80a0a41d314ddc5768a42915a69ca/runtime/compiler/runtime/J9Profiler.hpp#L660-L663

**Review by r30shah — CHANGES_REQUESTED** [review]
Some minor nit-picks I found while going through the changes. Also have you tested this change with the unit test or something. I think we should verify the block frequencies, let me know if you have already done it, if not, I can search through my old repo to find needed changes for debugging block frequencies.

File: runtime/compiler/runtime/J9Profiler.hpp
Comment (ashu-mehra) [reply runtime/compiler/runtime/J9Profiler.hpp:667]: @r30shah  If I understand this correctly the comment says if the entry has is low tagged it would have some integer value, otherwise it would be pointing to `TR_BitVector`.
`IS_VALID_BIT_VECTOR` returns true if the entry is not low tagged, which seems to be in-sync with the comments.
It is missing a NULL check, I can add that to the condition. Is there anything else you are pointing to?

**Comment by ashu-mehra** [comment]
> Also have you tested this change with the unit test or something.

Do we have unit tests for `BlockFrequency` class? I can update them if we already have and add the tests for the new methods.
If you can point to your changes for testing the code, I can try using that as well.

File: runtime/compiler/runtime/J9Profiler.hpp
Comment (r30shah) [reply runtime/compiler/runtime/J9Profiler.hpp:667]: Yeah, I think missing NULL check should be there. 
 What I meant originally is the name `IS_VALID_BIT_VECTOR` , the name seems more generic but the check is internal to BlockFrequencyInfo, so IMO it should either a name that justifies the internal use, or a private static method (To make code more readable) instead of the macro

**Comment by r30shah** [comment]
> Do we have unit tests for BlockFrequency class? I can update them if we already have and add the tests for the new methods.
If you can point to your changes for testing the code, I can try using that as well.

I was testing original changes with small unit test that exercises the profiling compilation. I think I can tweak it with the help of certain JIT options to run it in more controlled manner, and also a change to dump the Block Frequency Info. It definitely won't be the absolute unit test, but a frequencies reported in both configuration should be in same ball park. Let me try the changes out.

**Review by dsouzai — COMMENTED** [review]
Overall looks fine to me. However, I'll hold on approval / running tests till I get @r30shah's approval as he's the content expert here.

Also, you should update https://github.com/eclipse/openj9/issues/9974 as the `TR_Serializer` class doesn't exist anymore.

**Comment by ashu-mehra** [comment]
I have addressed the comments and squashed the commits as well.

**Comment by dsouzai** [comment]
@r30shah does everything look good to you?

**Review by r30shah — APPROVED** [review]
Changes looks good to me.

**Comment by dsouzai** [comment]
jenkins test sanity plinux jdk11

**Review by dsouzai — APPROVED** [review]

**Comment by dsouzai** [comment]
jenkins compile osx,aix,win

^ just to ensure no compile issues on non-gcc compilers.

**Comment by dsouzai** [comment]
jenkins compile osx,aix,win jdk11

**Comment by dsouzai** [comment]
jenkins compile osx,aix,win jdk11

**Pull request merged**
